### ISSUE TITLE
module weights don't update after copy.deepcopy

### ISSUE THREAD

**Issue by stsievert** [dnouri/skorch#317]
When a model is copied with `copy.deepcopy`, the weights are frozen and don't update when `fit` is called.

Here's an example:

``` python
from sklearn.datasets import make_regression
import torch.nn as nn
import torch
from skorch import NeuralNetRegressor
import copy
import toolz

def get_data(n, d):
    X, y = make_regression(n_features=d, n_samples=n)
    y -= y.min()
    y /= y.max()
    y = y.reshape(-1, 1).astype('float32')

    X = X.astype('float32')
    return X, y

X, y = get_data(1000, 28 * 28)

class Simple(nn.Module):
    def __init__(self):
        super().__init__()
        self.simple = nn.Sequential(
            nn.Linear(28 * 28, 1),
            nn.Sigmoid(),
        )
        
    def forward(self, x):
        return self.simple(x)
n1 = NeuralNetRegressor(
    module=Simple,
).initialize()

def check(net, name=''):
    p = toolz.first(net.module_.parameters()).detach().numpy()
    print(name + ' first params =', p.flat[:3])

def assert_all_params_equal(net1, net2):
    for p1, p2 in zip(net1.module_.parameters(), net2.module_.parameters()):
        assert p1.requires_grad
        assert p2.requires_grad
        assert torch.all(p1 == p2)
    
check(n1, name="n1")
n1.partial_fit(X, y)
check(n1, name="n1")

n2 = copy.deepcopy(n1)
check(n2, name="n2")
assert_all_params_equal(n1, n2)

n3 = n2.partial_fit(X, y[::-1])

check(n1, name="n1")
check(n2, name="n2")
check(n3, name="n3")

assert_all_params_equal(n2, n3)
assert_all_params_equal(n1, n2)
```

which, when run, outputs this:

```
n1 first params = [-0.03383204 -0.00945246  0.03146857]
  epoch    train_loss    valid_loss     dur
-------  ------------  ------------  ------
      1        0.0415        0.0529  0.0245
      2        0.0404        0.0523  0.0251
      3        0.0394        0.0517  0.0252
n1 first params = [-0.03349511 -0.00889289  0.02929294]
n2 first params = [-0.03349511 -0.00889289  0.02929294]
      4        0.0431        0.0487  0.0250
      5        0.0431        0.0487  0.0246
      6        0.0431        0.0487  0.0247
n1 first params = [-0.03349511 -0.00889289  0.02929294]
n2 first params = [-0.03349511 -0.00889289  0.02929294]
n3 first params = [-0.03349511 -0.00889289  0.02929294]
```

**Comment by stsievert** [comment]
I think this is happening because there's a disconnect between the gradients `optimizer` sees and the gradients the model weights calculate.

That is, with the same setup as above:

``` python
>>> def check(net, name='', grad=False):
...     p = toolz.first(net.module_.parameters()).detach().numpy()
...     if not grad:
...         print(name + ' first params =', p.flat[:3])
...     p = toolz.first(net.module_.parameters()).grad
...     if p is not None and grad:
...         p = p.numpy()
...         print(name + ' first params =', p.flat[:3])

>>> check(n1, name="n1", grad=True)
n1 first grad params = [-0.00635823 -0.01619332 -0.0140646 ]
>>> check(n2, name="n2", grad=True)
n2 first grad params = [ 0.04858593 -0.09615209  0.00394883]
```

The weights the optimizer sees are specified in `optimizer_.param_groups`: [torch/optim/sgd.py#L82]. Requiring the gradients for `loss.backward()` does nothing if the optimizer can't see them.

[torch/optim/sgd.py#L82]:https://github.com/pytorch/pytorch/blob/e449a27646f3b8b44e893e6be09606c4f24f7313/torch/optim/sgd.py#L82

**Comment by benjamin-work** [comment]
Thanks for reporting this bug. You are right, during the copying process, the parameters in the module and in the optimizer are copied separately, leading to the disconnect, which is why they don't update (though "frozen" doesn't quite fit it).

The easiest solution to me seems to be to add the following lines to the end of `__setstate__`:

```
        if self.initialized_:
            self.initialize_optimizer()
```

For your example code, the error is fixed by this snippet (btw. you don't need `toolz.first`, `next` does the same here). Could you confirm that this solves your initial problem?

The deeper problem, however, lies in the fact that we use `torch.save` on the `module_` and `optimizer_` here: https://github.com/dnouri/skorch/blob/be050a15534924e38eaa06480c7f071a2cce16c7/skorch/net.py#L1310-L1316

The reasoning here was to be able to load parameters to another device later in `__getstate__`, e.g. when you train on cuda and predict on cpu. We should, however, find a better solution to this, as with the current one, the parameters from `module_` and `optimizer_` are copied (leading to the bug).

Even with the proposed fix, we should still avoid the copy because it doubles the memory required for the state. Unfortunately, this will require a more specific solution instead of the more general use of `cuda_dependent_attributes_`, a solution that will fail if more attributes reference the parameters. But I guess that's still better than the current implementation.

**Comment by benjamin-work** [comment]
I digged a little deeper and it seems that a similar issue, the de-coupling of module and optimizer parameters, has tripped some pytorch users in the past:

https://discuss.pytorch.org/t/saving-and-loading-a-model-in-pytorch/2610

I tried to create a minimal example that demonstrates what is required to preserve the connection:

https://gist.github.com/benjamin-work/ca12831877c241935c224a34bcbda713

Critically, saving module and optimizer in [one data structure](https://gist.github.com/benjamin-work/ca12831877c241935c224a34bcbda713#file-pytorch_save_and_load-py-L72) instead of two solves the problem, which I admit was surprising to me. This, in conjunction with the use of `state_dict`, could point forward to a clean solution for the issue.

**Comment by stsievert** [comment]
> Could you confirm that this solves your initial problem?

The proposed solution solves my initial problem: the weights update. But the optimizer is reinitialized; I don't think any state is preserved. I've let that be; it seems like https://github.com/dnouri/skorch/issues/317#issuecomment-414629936 is a better solution.

> Critically, saving module and optimizer in one data structure ... use of state_dict, could point forward to a clean solution for the issue.

Perfect. See https://github.com/dnouri/skorch/pull/318. It does not use `state_dict`, but passes the test I wrote.
